FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt

```diff
@@ -57,12 +73,198 @@ object PgpMsg {
+    // or wrapped in PGP armor as plaintext packet without encrypting)
+    // also false when error happens.
+    val isEncrypted: Boolean = false,
+
+    // pgp messages may include original filename in them
+    val filename: String? = null,
+
+    // todo later - signature verification not supported on Android yet
+    val signature: Any? = null,
+
+    // provided if error happens
+    val error: DecryptionError? = null
+  ) {
+    companion object {
+      fun withError(
+        type: DecryptionErrorType,
+        message: String,
+        cause: Throwable? = null
+      ): DecryptionResult {
+        return DecryptionResult(error = DecryptionError(type, message, cause))
+      }
+
+      fun withCleartext(cleartext: ByteArrayOutputStream, signature: Any?): DecryptionResult {
+        return DecryptionResult(content = cleartext, signature = signature)
+      }
+
+      fun withDecrypted(content: ByteArrayOutputStream, filename: String?): DecryptionResult {
+        return DecryptionResult(content = content, isEncrypted = true, filename = filename)
+      }
+    }
+  }
+
+  data class KeyWithPassPhrase(
+    val keyRing: PGPSecretKeyRing,
+    val passphrase: Passphrase?
+  )
+
+  fun decrypt(
+    data: ByteArray,
+    keys: List<KeyWithPassPhrase>,
+    pgpPublicKeyRingCollection: PGPPublicKeyRingCollection? // for verification
+  ): DecryptionResult {
+    if (data.isEmpty()) {
+      return DecryptionResult.withError(DecryptionErrorType.FORMAT, "Can't decrypt empty message")
+    }
+
+    val chunk = data.copyOfRange(0, data.size.coerceAtMost(50)).toString(StandardCharsets.US_ASCII)
+    var input = data.inputStream() as InputStream
+
+    if (chunk.contains(ARMOR_HEADER_DICT[MsgBlock.Type.SIGNED_MSG]!!.begin)) {
+      val msg: PgpArmor.CleartextSignedMessage
+      try {
+        msg = PgpArmor.readSignedClearTextMessage(input)
+      } catch (ex: Exception) {
+        return if (
+          ex is PGPException && ex.message != null && ex.message == "Cleartext format error"
+        ) {
+          DecryptionResult.withError(
+            type = DecryptionErrorType.FORMAT,
+            message = ex.message!!,
+            cause = ex.cause
+          )
+        } else {
+          DecryptionResult.withError(
+            type = DecryptionErrorType.OTHER,
+            message = "Decode cleartext error",
+            cause = ex
+          )
+        }
+      }
+      return DecryptionResult.withCleartext(msg.content, msg.signature)
+    }
+
+    val isArmored = chunk.contains(ARMOR_HEADER_DICT[MsgBlock.Type.ENCRYPTED_MSG]!!.begin)
+    if (isArmored) input = ArmoredInputStream(input)
+
+    val keyList: List<PGPSecretKeyRing>
+    try {
+      keyList = keys.map {
+        if (it.passphrase != null) {
+          PgpKey.decryptKey(it.keyRing, it.passphrase)
+        } else {
+          if (KeyRingInfo(it.keyRing).isFullyDecrypted) {
+            throw PGPException("flowcrypt: need passphrase")
+          }
+          it.keyRing
```

Comment:
I think you meant to write `if(!KeyRingInfo(it.keyRing).isFullyDecrypted)`

However I would invert the logic - if it's already decrypted, then no need to worry about pass phrase.

```kt
if(KeyRingInfo(it.keyRing).isFullyDecrypted) {
  it.keyRing
} else if(it.passphrase == null) {
  throw PGPException("missing pass phrase to decrypt key")
} else {
  PgpKey.decryptKey(it.keyRing, it.passphrase) // may throw PGPException
}
```

Reply:
ok, will change this